proof/refine/ARM/KHeap_R.thy

```diff
@@ -3921,4 +3935,315 @@ lemma threadGet_getObject:
   apply (simp add: threadGet_def threadRead_def oliftM_def getObject_def[symmetric])
   done
 
+lemma obj_at'_typ_at'[elim!]:
+  "obj_at' (P :: ('a :: pspace_storable) \<Rightarrow> bool) p s \<Longrightarrow>
+   obj_at' (\<top> :: ('a :: pspace_storable) \<Rightarrow> bool) p s"
+  by (clarsimp simp: obj_at'_real_def ko_wp_at'_def)
+
+lemma sc_tcbs_of_pred_map_equiv:
+  "sc_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>x. scTCB x = Some t) p s = pred_map_eq t (scTCBs_of s) p"
```

Comment:
Nice, I think we've needed lemmas like these ones for a while, that link `obj_at'` and projections.

Reply:
I looked for this and could not find it, but it does exist in two places with reply objects. I like the direct equality that this pattern uses. We can discuss `pred_map` separately.

Reply:
Well, we have `pred_map_eq t (scTCBs_of s) p = scTCBs_of s p = Some t`. And we also know `obj_at' (\<lambda>x. scTCB x = Some t) p s = sc_at’ p s & scTCBs_of s p = Some t`. So, this equality is pretty immediate if we can assume `sc_at’ p s`.

The question is, why do we need to go via `pred_map_eq` when we could easily have equality for `scTCBs_of s p = Some t` directly?

Reply:
I just realised that this `obj_at (xxx)` pattern must have been copied from AInvs.

I would like you all to note the difference between `obj_at` and `obj_at’`. The latter asks for the alignment conditions, which we often do not need. If the information you actually need is only projections (object field values), we shouldn’t have `obj_at’`.

And, when we do need alignment conditions, it is better if we separate the `obj_at’` and the projection information because, by mixing those, we run the risk of having to prove alignment when we don’t need to.

The existing Refine proof is indeed a total mix of these, and we are forced to prove alignment again and again, which we could do without. So, for the new part of the proof we are adding, it would be good to be ensuring this distinction.

Reply:
In answer to your first question: `scTCBs_of s p = Some t` exposes heap evaluation to the simplifier via `opt_mapE`.
In this example
```
lemma test_lemma1:
  "⟦scTCBs_of s p = Some t; scs_of' s p = Some j⟧ ⟹ foo"
  apply (clarsimp)
  oops
```
the hypothesis will be immediately simplified to `scTCB sc = Some t; sc_of' v'a = Some sc; ksPSpace s p = Some v'a`. While this may provide detailed information about the state of the proof, it prevent the application of any drule or erule that would have matched on `scTCBs_of s p = Some t`. Yes, I can remove `opt_mapE` from [elim], but you could alternatively add `pred_map_eq` to [simp].

In this case, going via `pred_map_eq` preserves this common pattern and allows for repeated combination of other matching simp rules, erules and drules. I don't think using heaps directly will preserve this information.

Reply:
Addressing your second comment. I'm well aware of alignment information, it's somewhat patronising to have you explain it to me.

While this lemma is not optimal (I prefer the pattern you have provided in a separate comment), the entire point of the lemma is to remove `obj_at'` phrases from proofs in order to reason about heap values and ignore alignment. This lemma is close to the kind of replacement you highlighted elsewhere and so I'm not sure what point you're trying to make.

Reply:
>Yes, I can remove `opt_mapE` from [elim], but you could alternatively add `pred_map_eq` to [simp].

Removing `opt_mapE` from the rule set is what we have been doing for the whole of Refine proof so far. So, please stick with what we have been doing rather than introducing something different unless you are going to change the whole thing in that way.

Reply:
A further note, which supports for "hiding things behind definitions" is any proof of the form
```
lemma test_lemma2:
  "⟦sym_heap h1 h2'; foo⟧ ⟹ sym_heap h1 h2"
  apply_trace (clarsimp)
  oops
```
Any call to clarsimp (and not simp) will immediately introduce `\forall p p'.` into the context and leave a goal of the form `(h1 p = Some p') = (h2 p' = Some p)` which is not abbreviated to `sym_heap h1 h2` and so there is no way to use any rules or erules concerning `sym_heap`. In particular, `delta_sym_refs` is an elim rule of this form.

Reply:
> In this case, going via `pred_map_eq` preserves this common pattern and allows for repeated combination of other matching simp rules, erules and drules. I don't think using heaps directly will preserve this ~information~ pattern.

I don’t think I understand fully what you are trying to say here, but having this definition will leave us with numerous lemmas that we have written so far that are not in the `pred_map_eq` pattern that won’t match with then. Please consider that we have been doing fine with our approach so far.

Reply:
> Removing opt_mapE from the rule set is what we have been doing for the whole of Refine proof so far. So, please stick with what we have been doing rather than introducing something different unless you are going to change the whole thing in that way.
I'm really not introducing anything different Miki, there are 900+ instances of `pred_map` in Ainvs. I was never told that proofs in refine needed to avoid these patterns and this PR is the first time that we've had any conflict on the issue. If you want to make a new decision going forward, that's fine, but it's simply not true that I'm introducing something different.

Reply:
> A further note, which supports for "hiding things behind definitions" is any proof of the form
> 
> ```
> lemma test_lemma2:
>   "⟦sym_heap h1 h2'; foo⟧ ⟹ sym_heap h1 h2"
>   apply_trace (clarsimp)
>   oops
> ```
> 
> Any call to clarsimp (and not simp) will immediately introduce `\forall p p'.` into the context and leave a goal of the form `(h1 p = Some p') = (h2 p' = Some p)` which is not abbreviated to `sym_heap h1 h2` and so there is no way to use any rules or erules concerning `sym_heap`. In particular, `delta_sym_refs` is an elim rule which we might want to use.

For this, I have made this suggestion as a DM on mattermost to you:
"by the way, if all these argument can be resolved by having sym_heap as definition, please do consider that”
so please consider that - and I think that would be the only definition layer that we need.

Reply:
> For this, I have made this suggestion as a DM on mattermost to you:
"by the way, if all these argument can be resolved by having sym_heap as definition, please do consider that”
so please consider that - and I think that would be the only definition layer that we need.

Fair enough.

Reply:
I don't think this conversation is making much progress and personally I think there are benefits and negatives in both directions. Can I suggest that we get together offline to properly work through the options and decide on a path forwards?

Reply:
> > Removing opt_mapE from the rule set is what we have been doing for the whole of Refine proof so far. So, please stick with what we have been doing rather than introducing something different unless you are going to change the whole thing in that way.
> 
> I'm really not introducing anything different Miki, there are 900+ instances of `pred_map` in Ainvs. I was never told that proofs in refine needed to avoid these patterns and this PR is the first time that we've had any conflict on the issue. If you want to make a new decision going forward, that's fine, but it's simply not true that I'm introducing something different.

`pred_map` instances in AInvs have very little bearing on projections in Refine, I would think? And we are talking about `pred_map_eq`, right? We don’t have `pred_map_eq` in Refine, and we should keep it that way, is what I am saying.

Besides, just think about this. You are suggesting to introduce this `pred_map_eq` to package up `m x = Some y', and at the same time suggesting to put `pred_map_eq y m x = (m x = Some y)` in the simp set. I would think this is indicative that we could do without this roundtrip at all.

Reply:
> I don't think this conversation is making much progress and personally I think there are benefits and negatives in both directions. Can I suggest that we get together offline to properly work through the options and decide on a path forwards?

I’m happy to meet but not sure when...

Reply:
Sorry if I sounded patronising, that was not my intention. I just wanted to make sure that everyone is on the same page.

> While this lemma is not optimal (I prefer the pattern you have provided in a separate comment), the entire point of the lemma is to remove `obj_at'` phrases from proofs in order to reason about heap values and ignore alignment. This lemma is close to the kind of replacement you highlighted elsewhere and so I'm not sure what point you're trying to make.

My point is that this might not be the best pattern for Refine proofs. As I said earlier, it looks like this pattern is introduced in several places in the proof by instantiation, but this might not be the best approach in Refine.

Reply:
> Sorry if I sounded patronising, that was not my intention. I just wanted to make sure that everyone is on the same page.
> 
> > While this lemma is not optimal (I prefer the pattern you have provided in a separate comment), the entire point of the lemma is to remove `obj_at'` phrases from proofs in order to reason about heap values and ignore alignment. This lemma is close to the kind of replacement you highlighted elsewhere and so I'm not sure what point you're trying to make.
> 
> My point is that this might not be the best pattern for Refine proofs. As I said earlier, it looks like this pattern is introduced in several places in the proof by instantiation, but this might not be the best approach in Refine.

Thank you. Ok, if it's the `pred_map_eq` you'd prefer to avoid then I understand what you mean.